### ISSUE TITLE
Rework Scheduled Tasks And Use IHostedService Instead

### ISSUE THREAD

**Issue by toddca** [nopSolutions/nopCommerce#3859]
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/hosted-services?view=aspnetcore-5.0&tabs=visual-studio

**Comment by landlogicit** [comment]
Better change is HangFire

**Comment by a-patel** [comment]
+ HangFire

**Comment by miguelgimenez** [comment]
Hi,

We currently use Hangfire for all background processing in our nop plugins.

It has been working great with some sites running millions of tasks per month and it provides an excellent UI for monitoring tasks.

We could happily discuss ways of integrating what we have now into Nop

**Comment by AndreiMaz** [comment]
after some consideration we've decided not to implement this functionality

**Comment by RomanovM** [comment]
After completing #2287, we can again return to thinking about this.

**Comment by skoshelev** [comment]
Hangfire support only MS SQL Server only, we can't use it

**Comment by miguelgimenez** [comment]
Hi Sergey,

We have been using Hangfire to manage thousands of background tasks on over one hundred nopCommerce instances and it has allowed us to do thigs that are impossible with current task system.

Even though Hangfire natively only supports MS SQL as you mention it is extensible an there are storage implementations available both for My SQL and for PostgreSQL.

https://github.com/arnoldasgudas/Hangfire.MySqlStorage
https://github.com/frankhommers/Hangfire.PostgreSql

Having a robust background processing engine enables us to implement more advanced and complex functionality especially for integrations with third parties.

Implementing a robust background task an scheduling solution is a very complex undertaking and provably makes more sense to use something already built and tested.

We would be happy to share our code on how we integrated Hangfire engine and admin interface, if you want we can discuss through partner support.

Regards,

Miguel

**Comment by skoshelev** [comment]
Hi @miguelgimenez. Thank you very much for your willingness to help us, but our problem lies precisely in the fact that we need the support of all those databases that we support in the nopCommerce. Otherwise, it will be very strange to say to store owners: well, you are using MySQL, but also install Postgre for scheduled tasks started working. Of course, it is good to use the already tested functionality, but it should not limit the user and give him inconvenience

**Comment by miguelgimenez** [comment]
I completely understand the need to support at least all tree database engines supported by nopCommerce.

We are still using MS SQL for all your sites but will look closer at moving to one of the others in the future.

Still, my point is that there are plenty of Hangfire storage implementations available as you can see in https://www.hangfire.io/extensions.html including implementations for all database engines supported by nopCommerce at this time.

**Comment by skoshelev** [comment]
Hi @miguelgimenez.

Yes, there are implementations for all the databases we use. But, I noticed the provider for MySql (I did not look directly at everything), the branch with support for .net core has not been updated for three years, it has some problems with deadlocks and a queue that have not been resolved for more than a year ... We'll try to revise the standard TaskManager so that it can be extended or replaced from plugins, for those who want to use Hangfire, for example.
